### PR TITLE
fix(amplify-codegen-appsync-model-plugin): add ownerField to modelgen…

### PR DESCRIPTION
**Pull request by AaronZyLee**
*Issue #, if available:*
aws-amplify/amplify-js#5687
*Description of changes:*
`amplify codegen models` will generate `ownerField` automatically as an optional string type `String` in implicit cases and will not override the field if it is defined explicitly for other type in the schema, which allows user to use custom type such as `[String!]` etc. The changes apply to all the SDKs(js, android and ios)
### Sample input and output for implicit cases
Schema:
```
 type Post @model @auth(rules: [
  { allow: owner , ownerField: "customOwnerField1"},
  { allow: owner , ownerField: "customOwnerField2"},
  ]) {
   id: ID!
   name: String!
 } 
```
Java output:
```
/** This is an auto generated class representing the Post type in your schema. */
@SuppressWarnings("all")
@ModelConfig(pluralName = "Posts", authRules = {
  @AuthRule(allow = AuthStrategy.OWNER, ownerField = "customOwnerField1", identityClaim = "cognito:username", operations = { ModelOperation.CREATE, ModelOperation.UPDATE, ModelOperation.DELETE, ModelOperation.READ }),
  @AuthRule(allow = AuthStrategy.OWNER, ownerField = "customOwnerField2", identityClaim = "cognito:username", operations = { ModelOperation.CREATE, ModelOperation.UPDATE, ModelOperation.DELETE, ModelOperation.READ })
})
public final class Post implements Model {
  public static final QueryField ID = field("id");
  public static final QueryField NAME = field("name");
  public static final QueryField CUSTOM_OWNER_FIELD1 = field("customOwnerField1");
  public static final QueryField CUSTOM_OWNER_FIELD2 = field("customOwnerField2");
  private final @ModelField(targetType="ID", isRequired = true) String id;
  private final @ModelField(targetType="String", isRequired = true) String name;
  private final @ModelField(targetType="String") String customOwnerField1;
  private final @ModelField(targetType="String") String customOwnerField2;
  public String getId() {
      return id;
  }
  
  public String getName() {
      return name;
  }
  
  public String getCustomOwnerField1() {
      return customOwnerField1;
  }
  
  public String getCustomOwnerField2() {
      return customOwnerField2;
  }
//...
//Constructors, getters and builders also get updated. See details in test snapshots.
```
Swift output:
```
// swiftlint:disable all
import Amplify
import Foundation

extension Post {
  // MARK: - CodingKeys 
   public enum CodingKeys: String, ModelKey {
    case id
    case name
    case customOwnerField1
    case customOwnerField2
  }
  
  public static let keys = CodingKeys.self
  //  MARK: - ModelSchema 
  
  public static let schema = defineSchema { model in
    let post = Post.keys
    
    model.authRules = [
      rule(allow: .owner, ownerField: "customOwnerField1", identityClaim: "cognito:username", operations: [.create, .update, .delete, .read]),
      rule(allow: .owner, ownerField: "customOwnerField2", identityClaim: "cognito:username", operations: [.create, .update, .delete, .read])
    ]
    
    model.pluralName = "Posts"
    
    model.fields(
      .id(),
      .field(post.name, is: .required, ofType: .string),
      .field(post.customOwnerField1, is: .optional, ofType: .string),
      .field(post.customOwnerField2, is: .optional, ofType: .string)
    )
    }
}
```
### Sample input and output(swift) for user-defined-type ownerField
Schema:
```
type Post
  @model
  @auth(rules: [
    { allow: owner, ownerField: "author" }, 
    { allow: owner, ownerField: "editors", operations: [update, read] }
  ]) {
  id: ID!
  title: String!
  author: String!
  editors: [String!]
}
```
Output:
```
"// swiftlint:disable all
import Amplify
import Foundation

extension Post {
  // MARK: - CodingKeys 
    public enum CodingKeys: String, ModelKey {
    case id
    case title
    case author
    case editors
  }
  
  public static let keys = CodingKeys.self
  //  MARK: - ModelSchema 
  
  public static let schema = defineSchema { model in
    let post = Post.keys
    
    model.authRules = [
      rule(allow: .owner, ownerField: \\"author\\", identityClaim: \\"cognito:username\\", operations: [.create, .update, .delete, .read]),
      rule(allow: .owner, ownerField: \\"editors\\", identityClaim: \\"cognito:username\\", operations: [.update, .read])
    ]
    
    model.pluralName = \\"Posts\\"
    
    model.fields(
      .id(),
      .field(post.title, is: .required, ofType: .string),
      .field(post.author, is: .required, ofType: .string),
      .field(post.editors, is: .required, ofType: .embeddedCollection(of: String.self))
    )
    }
}"
```
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by richardmcclellan — APPROVED** [review]
:shipit:

**Review by wooj2 — COMMENTED** [review]
in looking at your example, i noticed that your test input is:
```
  { allow: owner , ownerField: "customOwnerField1"},
  { allow: owner , ownerField: "customOwnerField2"},
```
Is this a valid use case?  I have never seen a case where we would have two different custom owner fields.

**Comment by richardmcclellan** [comment]
> in looking at your example, i noticed that your test input is:
> 
> ```
>   { allow: owner , ownerField: "customOwnerField1"},
>   { allow: owner , ownerField: "customOwnerField2"},
> ```
> 
> Is this a valid use case? I have never seen a case where we would have two different custom owner fields.

[The docs](https://docs.amplify.aws/cli/graphql-transformer/directives#multiple-authorization-rules) describe one such use case where an owner should be able to do perform all operations (create, read, update, delete), but a list of `editors` should only be able to read and update.

**Comment by wooj2** [comment]
> > in looking at your example, i noticed that your test input is:
> > ```
> >   { allow: owner , ownerField: "customOwnerField1"},
> >   { allow: owner , ownerField: "customOwnerField2"},
> > ```
> > 
> > 
> > Is this a valid use case? I have never seen a case where we would have two different custom owner fields.
> 
> [The docs](https://docs.amplify.aws/cli/graphql-transformer/directives#multiple-authorization-rules) describe one such use case where an owner should be able to perform all operations (create, read, update, delete), but a list of `editors` should only be able to read and update.

Seems like we should test that case instead?

**Comment by zeynepsu** [comment]
Should there be a corresponding test for swift visitor as well?

**Comment by AaronZyLee** [comment]
> > > in looking at your example, i noticed that your test input is:
> > > ```
> > >   { allow: owner , ownerField: "customOwnerField1"},
> > >   { allow: owner , ownerField: "customOwnerField2"},
> > > ```
> > > 
> > > 
> > > Is this a valid use case? I have never seen a case where we would have two different custom owner fields.
> > 
> > 
> > [The docs](https://docs.amplify.aws/cli/graphql-transformer/directives#multiple-authorization-rules) describe one such use case where an owner should be able to perform all operations (create, read, update, delete), but a list of `editors` should only be able to read and update.
> 
> Seems like we should test that case instead?

I include the test cases in the new commit for swift tests.

File: packages/amplify-codegen-appsync-model-plugin/src/utils/process-auth.ts [hunk 1]
```diff
@@ -92,31 +92,4 @@ export function processAuthDirective(directives: CodeGenDirectives): AuthDirecti
       },
     };
   });
-}
-
-export function getOwnerAuthRules(modelObj: CodeGenModel): AuthRule[] {
```
Comment (wooj2) [inline packages/amplify-codegen-appsync-model-plugin/src/utils/process-auth.ts:97]: I'm not familiar with this code, but it seems like we are removing `getOwnerAuthRules` (and getOwnerFieldName, etc..) from everywhere.  Can anyone verify that this is the right thing to do?

File: packages/amplify-codegen-appsync-model-plugin/src/__tests__/visitors/appsync-swift-visitor.test.ts [hunk 2]
```diff
@@ -1160,6 +1160,145 @@ describe('AppSyncSwiftVisitor', () => {
...
+        const schema = /* GraphQL */ `
+          type Post
+            @model
+            @auth(rules: [{ allow: owner, ownerField: "author" }, { allow: owner, ownerField: "editors", operations: [update, read] }]) {
+            id: ID!
+            title: String!
+            author: String!
+            editors: [String!]
+          }
+        `;
```
[the remark this answers is not in the document]
Comment (AaronZyLee) [reply packages/amplify-codegen-appsync-model-plugin/src/__tests__/visitors/appsync-swift-visitor.test.ts:1262]: The temporary solution for missing ownerField type is optional string, which is `String`.

[the remark this answers is not in the document]
Comment (AaronZyLee) [reply packages/amplify-codegen-appsync-model-plugin/src/__tests__/visitors/appsync-swift-visitor.test.ts:1262]: I am not sure if these tests are necessary. I think the leaving out variable test is already covered by the test case `it('should add the custom ownerField to model generation automatically if not provided in schema')`, which proves that a missing field will be added with default type.

**Review by yuth — CHANGES_REQUESTED** [review]
Could you fix the LGTM errors. Other than that it looks good

**Review by yuth — APPROVED** [review]

**Pull request merged**